File: frontend/src/services/group-service.ts
import { httpClient } from '../lib';
import type {
  GroupResponse,
  GroupDetailResponse,
  GroupMemberResponse,
  StudentSearchItem,
  CreateGroupPayload,
  UpdateGroupPayload,
  AddMembersPayload,
  PagedResponse,
} from '../types';

type SpringPage<T> = {
  content: T[];
  totalElements: number;
  totalPages: number;
  number: number;
  size: number;
};

export async function getGroups(params?: {
  courseId?: string;
  keyword?: string;
  page?: number;
  size?: number;
}) {
  const response = await httpClient.get<SpringPage<GroupResponse>>(
    '/instructor/groups',
    { params: { ...params, size: params?.size ?? 100 } },
  );
  return response.data;
}

export async function getGroupDetail(groupId: string) {
  const response = await httpClient.get<GroupDetailResponse>(
    `/instructor/groups/${groupId}`,
  );
  return response.data;
}

export async function createGroup(payload: CreateGroupPayload) {
  const response = await httpClient.post<GroupResponse>(
    '/instructor/groups',
    payload,
  );
  return response.data;
}

export async function updateGroup(groupId: string, payload: UpdateGroupPayload) {
  const response = await httpClient.put<GroupResponse>(
    `/instructor/groups/${groupId}`,
    payload,
  );
  return response.data;
}

export async function deleteGroup(groupId: string) {
  await httpClient.delete(`/instructor/groups/${groupId}`);
}

export async function getUnassignedStudents(courseId?: string) {
  const response = await httpClient.get<StudentSearchItem[]>(
    '/instructor/groups/students/unassigned',
    { params: { ...(courseId ? { courseId } : {}) } },
  );
  return response.data;
}

export async function addGroupMembers(groupId: string, payload: AddMembersPayload) {
  const response = await httpClient.post<GroupMemberResponse[]>(
    `/instructor/groups/${groupId}/members`,
    payload,
  );
  return response.data;
}

export async function removeGroupMember(groupId: string, studentId: string) {
  await httpClient.delete(`/instructor/groups/${groupId}/members/${studentId}`);
}

export async function getStudentGroups() {
  const response = await httpClient.get<GroupResponse[]>(
    '/student/groups',
  );
  return response.data;
}

export async function getStudentGroupDetail(groupId: string) {
  const response = await httpClient.get<GroupDetailResponse>(
    `/student/groups/${groupId}`,
  );
  return response.data;
}

File: backend/src/main/java/project/lms_rikkei_edu/modules/group/service/GroupService.java
package project.lms_rikkei_edu.modules.group.service;

import jakarta.validation.Valid;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import project.lms_rikkei_edu.modules.group.dto.request.AddGroupMembersRequest;
import project.lms_rikkei_edu.modules.group.dto.request.CreateGroupRequest;
import project.lms_rikkei_edu.modules.group.dto.request.UpdateGroupRequest;
import project.lms_rikkei_edu.modules.group.dto.response.GroupDetailResponse;
import project.lms_rikkei_edu.modules.group.dto.response.GroupMemberResponse;
import project.lms_rikkei_edu.modules.group.dto.response.GroupResponse;
import project.lms_rikkei_edu.modules.group.dto.response.StudentSearchResponse;

import java.util.List;
import java.util.UUID;

public interface GroupService {

    Page<GroupResponse> getGroups(UUID courseId, String keyword, Pageable pageable);

    GroupDetailResponse getGroupDetail(UUID groupId);

    GroupResponse createGroup(@Valid CreateGroupRequest request);

    GroupResponse updateGroup(UUID groupId, @Valid UpdateGroupRequest request);

    void deleteGroup(UUID groupId);

    List<GroupMemberResponse> addMembers(UUID groupId, @Valid AddGroupMembersRequest request);

    void removeMember(UUID groupId, UUID studentId);

    List<StudentSearchResponse> searchStudentsByEmail(String email);

    List<StudentSearchResponse> getUnassignedStudents(UUID courseId);

    GroupDetailResponse getStudentGroupDetail(UUID groupId);

    List<GroupResponse> getStudentGroups();
}
